subprojects/core/src/integTest/groovy/org/gradle/api/provider/ProviderIntegrationTest.groovy

```diff
@@ -0,0 +1,381 @@
+                    enabled.get()
+                }
+                
+                void setEnabled(Provider<Boolean> enabled) {
+                    this.enabled = enabled
+                }
+                
+                @OutputFiles
+                FileCollection getOutputFiles() {
+                    outputFiles.get()
+                }
+
+                void setOutputFiles(Provider<FileCollection> outputFiles) {
+                    this.outputFiles = outputFiles
+                }
+                
+                @TaskAction
+                void resolveValue() {
+                    if (getEnabled()) {
+                        getOutputFiles().each {
+                            it.text = '$OUTPUT_FILE_CONTENT'
+                        }
+                    }
+                }
+            }
+        """
+
+        when:
+        succeeds('myTask')
+
+        then:
+        !defaultOutputFile.exists()
+    }
+
+    @Unroll
+    def "can lazily map extension property value to task property with #mechanism"() {
+        given:
+        buildFile << pluginWithExtensionMapping { taskConfiguration }
+        buildFile << customGroovyBasedTaskType()
+
+        when:
+        succeeds('myTask')
+
+        then:
+        !defaultOutputFile.exists()
+        customOutputFile.isFile()
+        customOutputFile.text == OUTPUT_FILE_CONTENT
+
+        where:
+        mechanism            | taskConfiguration
+        'convention mapping' | taskConfiguredWithConventionMapping()
+        'provider'           | taskConfiguredWithProvider()
+    }
+
+    def "can use provider to define a task dependency"() {
+        given:
+        buildFile << """
+            task producer {
+                ext.destFile = file('test.txt')
+                outputs.file destFile
+
+                doLast {
+                    destFile << '$OUTPUT_FILE_CONTENT'
+                }
+            }
+
+            def targetFile = provider { producer.destFile }
+            
+            targetFile.builtBy producer
+
+            task consumer {
+                dependsOn targetFile
+
+                doLast {
+                    println targetFile.get().text
+                }
+            }
+        """
+
+        when:
+        succeeds('consumer')
+
+        then:
+        executedTasks.containsAll(':producer', ':consumer')
+        outputContains(OUTPUT_FILE_CONTENT)
+    }
+
+    def "can use provider type to infer task dependency"() {
+        given:
+        buildFile << """
+            task producer(type: Producer) {
+                text = '$OUTPUT_FILE_CONTENT'
+                outputFiles = provider { files("\$buildDir/helloWorld.txt") }
+            }
+
+            task consumer(type: Consumer) {
+                inputFiles = producer.outputs.files
+            }
+
+            class Producer extends DefaultTask {
+                @Input
+                String text
+
+                Provider<FileCollection> outputFiles
```

Comment:
A couple of thoughts on `FileCollection` is-a `Provider`:

- If we'd go for it I believe we'd have to implement `Provider` for a `FileCollection` and `FileTree` separately. I don't think setting a `Set<File>` would work well for a `FileTree`. I'd rather see us providing a `dir` for the `set` method. I'd make the following changes:
    - `public interface FileCollection extends PropertyState<Iterable<File>>`
    - `public interface ConfigurableFileTree extends PropertyState<File>`
- The creation of a `FileCollection` or `FileTree` would happen as usual (via the methods on `Project`). The reason I am saying that is that `final ConfigurableFileCollection files = property(ConfigurableFileCollection)` probably wouldn't work as the generic type of the `Provider` would be `ConfigurableFileCollection` and not `Set<File>`. Examples:
    - `ConfigurableFileCollection outputFiles = project.files()`
    - `ConfigurableFileTree outputDir = project.fileTree('someDir')`
- I am still not convinced that users would adopt the pattern that a `FileCollection` is-a `Provider`. Most likely I'd expect users to still create `project.property(FileCollection)` and simply deal with a `PropertyState<FileCollection`. Using two different approaches for solving the same problem is probably going to be confusing. Most users normally don't have that intricate knowledge and want a single way of doing things. See the following example for illustration.

```
class MyTask extends DefaultTask {
    private final PropertyState<Boolean> enabled = project.property(Boolean)
    private final ConfigurableFileCollection outputFiles = project.files()
    
    MyTask() {
        enabled.set(false)
    }

    @Input
    boolean getEnabled() {
        enabled.get()
    }
    
    void setEnabled(boolean enabled) {
        this.enabled.set(enabled)
    }
    
    @OutputFiles
    FileCollection getOutputFiles() {
        outputFiles
    }

    void setOutputFiles(FileCollection outputFiles) {
        this.outputFiles.set(outputFiles)
    }
```

Any thoughts on my points?